### ISSUE TITLE
[chip-test] chip_sw_alert_handler_lpg_clock_off

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#14126]
### Test point name

[chip_sw_alert_handler_lpg_clock_off](https://github.com/lowRISC/opentitan/blob/b65e2705eb4ee814b46939d44f84d1adfee133c3/hw/top_earlgrey/data/chip_testplan.hjson#L1302)

### Host side component

None Required

### Contact person

@msfschaffner

### Checklist

Please fill out this checklist as items are completed. Link to PRs and issues as appropriate.
- [x] Check if existing test covers most or all of this testpoint (if so, either extend said test to cover all points, or skip the next 3 checkboxes)
- [x] Device-side (C) component developed
- [x] Bazel build rules developed
- [x] Host-side component developed
- [x] HJSON test plan updated with test name (so it shows up in the dashboard)
- [x] Test added to dvsim nightly regression (and passing at time of checking)

**Comment by bilgiday** [comment]
@msfschaffner, @cindychip,

For the first part of the test, I developed an initial test code based on the alert_handler_ping_timeout_test and created a draft PR #14858 so that DV team can reproduce the results below. 
- In the code, I turn off the hintable and gateable clocks one by one, and check if a ping-fail alert is triggered by the corresponding alert inputs.
   - If a clock is not gated, the ping-fail alert must be triggered
   - If a clock is gated/turned off, no ping fail alert must be triggered.

Wen I run the code on a CW310, I encounter a few cases that do not fully match with my expectations/understanding. I would like to share them with you to keep track of the discussion and to see if I am missing anything. @cindychip is checking the DV waveforms to figure out the root cause:

1. if I set the ping_timeout value  to 1 (`line 141`), I always get a ping-fail error independent of the clock states. Is this an intende/expected behavior?

2. If I set the ping_timeout value to 2, the ping-fail mechanism works as expected for the `gateable clocks`: if a clock is gated, no ping-fail alert is triggered. If a clock is NOT gated, a ping-fail alert is triggered. `NO PROBLEM HERE`

3. If I set the ping_timeout value to 2, I get the following behavior for the `hintable clocks`:
   - `Turning off the aes clock` affects ping-timeout behavior for all the alerts coming from aes, hmac, kmac, and otbn: If aes clock is gated, NONE of the alerts coming from these IPs triggers the ping-fail timeout alert
   - `Turning off the hmac/kmac/otbn clocks` seems to have no affect on the ping-timeout behavior: If any of these clocks is gated, ALL of the alerts coming from these IPs trigger the ping-fail timeout alert independent of the clock state.

**Comment by tjaychen** [comment]
@jeoongp 
let me help double check something. 
This area could be a bit weird depending on how the FPGA CG cells were hooked up.

**Comment by msfschaffner** [comment]
Hey @bilgiday, 

I am not 100% sure the test as you describe it matches with the description in the testplan. 
Let me clarify: we should NEVER see a ping fail alert if the alert handler is set up correctly, unless the alert channel signals are being tampered with. So for this particular test, this means that
- We should never get a ping failure alert, no matter whether a clock is gated or not.
- If a clock is NOT gated, the affected alert sender on that clock should work and any alert events (e.g. triggered via an alert test register) should be successfully transmitted.

RE question 1) above:
Setting the timeout to a value that is too small may lead to false positives, since the ping handshake takes time to complete. The time that it takes to complete the handshake is determined by the clock frequency of both sender and receiver domains. 
As mentioned in the docs [here](https://docs.opentitan.org/hw/top_earlgrey/ip_autogen/alert_handler/doc/index.html#Reg_ping_timeout_cyc_shadowed), it is recommended to set that timeout to the equivalent of 256 cycles of the slowest alert sender clock domain in the system. In our case, this means that the value should be set to 256, since the alert handler is running on IO_DIV4, which is the slowest clock in the system to which alert senders are attached to (there is also the AON clock, but it does not directly drive alert senders).

RE question 3) above:
The issue that @tjaychen mentions may exist (i.e., there may be clock domains which are not fully gated on the FPGA) - but I don't think it should create any ping fail alerts, since gated domains disable the alert channels via the LPG functionality in the alert handler. So it may be that this issue only occurs due to the timeout being set to a value that is too low.

**Comment by jeoongp** [comment]
Hey @bilgiday, is this issue occurring only on FPGA?

**Comment by bilgiday** [comment]
@msfschaffner, thanks for the feedback and the clarification. I was using a low-timeout value to create a negative test case to verify my own understanding of the IP. I will follow your description and use `ping_timeout=256` in the final code. 

@jeoongp, I am getting different behavior behavior from FPGA and Verilator. I need to do a few more tests to figure out what is going on.

**Comment by msfschaffner** [comment]
Ah I see. That makes sense yes.

**Comment by cindychip** [comment]
@bilgiday thanks for the detailed report.
I commented here in your PR: https://github.com/lowRISC/opentitan/pull/14858#discussion_r969047700
I wonder if that helps clear some error?

**Comment by bilgiday** [comment]
@cindychip, thanks for the suggestion, I removed all the local alerts except `kDifAlertHandlerLocalAlertAlertPingFail` but it did not change the behavior.

@jeoongp, I modified a few things to make my Verilator test more reliable: The C code needs to wait for enough time before checking the alert status. To shorten the wait time in Verilator, I manually modified `PING_CNT_DW` parameter to `4` in `hw/top_earlgrey/ip_autogen/alert_handler/rtl/alert_handler_reg_pkg.sv` and in `hw/top_earlgrey/ip_autogen/alert_handler/rtl/alert_handler.hjson`. Then, I confirmed that I get the same hinteable-clock behavior (observation 3 in [the comment 2](https://github.com/lowRISC/opentitan/issues/14126#issuecomment-1243420049)) on FPGA and Verilator.

To keep things simple, here is my proposal regarding how to proceed:

1. First, finish and submit the test based on the testplan's/Michael's definition.
2. Then, create a separate issue regarding the transactional-clock behavior. We can then decide if it requires more investigation, and based on our decision, I can put more focused effort on this one.

Does this sound good to everyone?

**Comment by jeoongp** [comment]
@bilgiday thanks! Can you provide a command to run observation 3 on FPGA and Verilator?

**Comment by jeoongp** [comment]
I modified PING_CNT_DW to 4 and ran the below command in #14858.
./util/dvsim/dvsim.py ./hw/top_earlgrey/dv/chip_sim_cfg.hjson -i chip_sw_alert_handler_lpg_clkoff 

Errors occurred as follows.
UVM_ERROR @ 12118.828277 us: (chip_sw_base_vseq.sv:223) uvm_test_top.env.virtual_sequencer [uvm_test_top.env.virtual_sequencer.chip_sw_base_vseq] SW TEST TIMED OUT. STATE: SwTestStatusInTest, TIMEOUT = 12000000 ns
UVM_ERROR @ 12118.828277 us: (tlul_assert.sv:296) [ASSERT FAILED] noOutstandingReqsAtEndOfSim_A

Is it the same results as your local run?

**Comment by bilgiday** [comment]
@cindychip,@msf, I created another version of the test, which is more parameterized and almost ready to be submitted. Before finalizing, I need your input:

- I implemented 3 test phases just in case: 2 in-spec phases (ping_timeout=256, fatal alerts) and 1 out-of-spec test (ping_timeout=2). I can completely remove the out-of-spec test if it doesn't improve the coverage, or I can move it into another "draft PR" if we would like to keep the code just in case. Please advise me. 

- Could you please take a look at the definitions of the test phases within the C code to see if they look good to you?

**Comment by msfschaffner** [comment]
Thanks @bilgiday!

RE the out-of-spec test: do you have an idea how much longer it takes to simulate with that phase included? 
it does not hurt keep it, but we should check whether the test simulation is reasonable.

RE the alert phases - see comments on the PR.

**Comment by tjaychen** [comment]
@bilgiday i'll close this down for now since #14858 is merged.
Feel free to re-open if anything pops up.

**Comment by cindychip** [comment]
Yes please! Sorry about that @tjaychen. We have one last piece that did not finish in the current test -
we need a separate DV testbench to force a fatal error and turn off the clock. Then check when the clock turned back on, the fatal alert keeps firing.

**Comment by tjaychen** [comment]
sounds good!

**Comment by engdoreis** [comment]
This test is passing since Oct 21th. Can we close this issue @tjaychen @msfschaffner?

|    | Tests                            |   latest |   2022.10.23 |   2022.10.22 |   2022.10.21 | 2022.10.17/2   | 2022.10.17/1   | 2022.10.15   | 2022.10.14/2   | 2022.10.14/1   | 2022.10.12   | Suite          |
|---:|:---------------------------------|---------:|-------------:|-------------:|-------------:|:---------------|:---------------|:-------------|:---------------|:---------------|:-------------|:---------------|
| 28 | chip_sw_alert_handler_lpg_clkoff |      100 |          100 |          100 |          100 | --             | --             | --           | --             | --             | --           | Unmapped tests |

**Comment by tjaychen** [comment]
okay with me. @bilgiday should probably give the final answer on whether he expects things to be okay now though.

**Comment by msfschaffner** [comment]
I think this is left open in order to track another, SV-only component of this test point that has not been developed yet. See comments further above.

**Comment by sriyerg** [comment]
> I think this is left open in order to track another, SV-only component of this test point that has not been developed yet. See comments further above.

SV portion completed in #16193.

**Comment by sriyerg** [comment]
The next steps for this test is documented in comment https://github.com/lowRISC/opentitan/issues/16190#issuecomment-1311261407, in issue #16190. Now that PR #16193 is closed, I am reopening this issue to track progress on those next steps. Reassigning to @bilgiday to take over and complete those steps.

**Comment by bilgiday** [comment]
@johngt, @johannheyszl The PR is merged, the test is passing on my local DV and FPGA environment. Waiting for nightly regression result to close the issue.

**Comment by bilgiday** [comment]
After #16305, the test has been passing the last 2 nightly regressions. We are tracking  the enhancement for M3 in #16336. 

Could we close the issue for M2 if everyone is okay?

**Comment by msfschaffner** [comment]
For initial development of a test, we require `Test added to dvsim nightly regression (and passing at time of checking)`. Hence it is ok to close this.

If it breaks in a future nightly regression, we will open a test-triage issue.
